### PR TITLE
Fixed duplication of falling blocks

### PR DESCRIPTION
**Pull request by SleepyScallop**
You could right click these items onto the bottom of plants and duplicate them, as they would stay in place and get dropped as well causing 2 items to appear. This fixes that.

### CODE REVIEW

File: src/main/java/io/github/thebusybiscuit/exoticgarden/listeners/PlantsListener.java [hunk 1]
```diff
@@ -353,11 +353,41 @@ public void onDecay(LeavesDecayEvent e) {
...
+    Material.GRAY_CONCRETE_POWDER,
+    Material.LIGHT_GRAY_CONCRETE_POWDER,
+    Material.CYAN_CONCRETE_POWDER,
+    Material.PURPLE_CONCRETE_POWDER,
+    Material.BLUE_CONCRETE_POWDER,
+    Material.BROWN_CONCRETE_POWDER,
+    Material.GREEN_CONCRETE_POWDER,
+    Material.RED_CONCRETE_POWDER,
+    Material.BLACK_CONCRETE_POWDER,
+    };
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/exoticgarden/listeners/PlantsListener.java]: This is what Tags are for.
I believe Slimefun even has one built-in too.

https://github.com/Slimefun/Slimefun4/blob/master/src/main/java/io/github/thebusybiscuit/slimefun4/utils/tags/SlimefunTag.java#L254

File: src/main/java/io/github/thebusybiscuit/exoticgarden/listeners/PlantsListener.java [hunk 2]
```diff
@@ -353,11 +353,41 @@ public void onDecay(LeavesDecayEvent e) {
...
+    
     @EventHandler
     public void onInteract(PlayerInteractEvent e) {
         if (e.getAction() != Action.RIGHT_CLICK_BLOCK) return;
         if (e.getHand() != EquipmentSlot.HAND) return;
         if (e.getPlayer().isSneaking()) return;
+        for(Material mat : blockedMaterials) {
+        if(mat == e.getPlayer().getInventory().getItemInMainHand().getType() || mat == e.getPlayer().getInventory().getItemInOffHand().getType()) return;
+        }
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/exoticgarden/listeners/PlantsListener.java]: Also, formatting and please use variables for long instructions

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]

**Comment by TheBusyBiscuit** [comment]
Also do you have a video of this happening?
I am not quite sure I can follow as to what ther underlying cause is.

**Comment by Sleepyflea** [comment]
> Also do you have a video of this happening? I am not quite sure I can follow as to what ther underlying cause is.

No, just use the latest version, right click a falling block onto a fruit of a tree for example a coconut tree.
Somtimes it will fall, other times it will stay put and duplicate the item. I will be looking into tags shortly.

**Comment by SleepyScallop** [comment]
> > Also do you have a video of this happening? I am not quite sure I can follow as to what ther underlying cause is.
> 
> No, just use the latest version, right click a falling block onto the bottom of  a fruit of a tree for example a coconut tree. Somtimes it will fall, other times it will stay put and duplicate the item. I will be looking into tags shortly.

**Comment by SleepyScallop** [comment]
Should be better now.

File: src/main/java/io/github/thebusybiscuit/exoticgarden/listeners/PlantsListener.java (hunk 1 above)
Comment (SleepyScallop) [reply src/main/java/io/github/thebusybiscuit/exoticgarden/listeners/PlantsListener.java]: Fixed!

File: src/main/java/io/github/thebusybiscuit/exoticgarden/listeners/PlantsListener.java (hunk 2 above)
Comment (SleepyScallop) [reply src/main/java/io/github/thebusybiscuit/exoticgarden/listeners/PlantsListener.java]: > Also, formatting and please use variables for long instructions

Fixed

**Comment by SleepyScallop** [comment]
> Also do you have a video of this happening? I am not quite sure I can follow as to what ther underlying cause is.

Heres your video
https://www.youtube.com/watch?v=caSLmVYWN4c

This code i just pushed fixes it.

File: src/main/java/io/github/thebusybiscuit/exoticgarden/listeners/PlantsListener.java [hunk 3]
```diff
@@ -353,11 +357,18 @@ public void onDecay(LeavesDecayEvent e) {
...
+    
     @EventHandler
     public void onInteract(PlayerInteractEvent e) {
         if (e.getAction() != Action.RIGHT_CLICK_BLOCK) return;
         if (e.getHand() != EquipmentSlot.HAND) return;
         if (e.getPlayer().isSneaking()) return;
+        for (SlimefunTag tag : valuesCache) {
+            nameLookup.put(tag.name(), SlimefunTag.GRAVITY_AFFECTED_BLOCKS);
+            if(tag.isTagged(mainHand) || tag.isTagged(offHand)) return;
+            }
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/exoticgarden/listeners/PlantsListener.java:371]: What the f*** are you doing here?
This is not how this works...

File: src/main/java/io/github/thebusybiscuit/exoticgarden/listeners/PlantsListener.java [hunk 4]
```diff
@@ -353,11 +357,18 @@ public void onDecay(LeavesDecayEvent e) {
         }
     }
 
+    private static final Map<String, SlimefunTag> nameLookup = new HashMap<>();
+    private static final SlimefunTag[] valuesCache = values();
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/exoticgarden/listeners/PlantsListener.java:361]: This is not needed.

File: src/main/java/io/github/thebusybiscuit/exoticgarden/listeners/PlantsListener.java [hunk 5]
```diff
@@ -47,6 +49,8 @@
 import io.github.thebusybiscuit.slimefun4.libraries.dough.skins.PlayerSkin;
 import io.github.thebusybiscuit.slimefun4.libraries.paperlib.PaperLib;
 import me.mrCookieSlime.Slimefun.api.BlockStorage;
+import io.github.thebusybiscuit.slimefun4.utils.tags.SlimefunTag;
+import static io.github.thebusybiscuit.slimefun4.utils.tags.SlimefunTag.values;
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/exoticgarden/listeners/PlantsListener.java:53]: This static import is not needed here...

File: src/main/java/io/github/thebusybiscuit/exoticgarden/listeners/PlantsListener.java (hunk 3 above)
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/exoticgarden/listeners/PlantsListener.java:371]: Also formatting

Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/exoticgarden/listeners/PlantsListener.java:371]: (Pseudo code) This is what you are supposed to do.
```java
        if (SlimefunTagGRAVITY_AFFECTED_BLOCKS.isTagged(...)) {
            return;
        }
```

File: src/main/java/io/github/thebusybiscuit/exoticgarden/listeners/PlantsListener.java [hunk 6]
```diff
@@ -353,11 +357,18 @@ public void onDecay(LeavesDecayEvent e) {
...
+    private static final SlimefunTag[] valuesCache = values();
+    
     @EventHandler
     public void onInteract(PlayerInteractEvent e) {
         if (e.getAction() != Action.RIGHT_CLICK_BLOCK) return;
         if (e.getHand() != EquipmentSlot.HAND) return;
         if (e.getPlayer().isSneaking()) return;
+        for (SlimefunTag tag : valuesCache) {
+            nameLookup.put(tag.name(), SlimefunTag.GRAVITY_AFFECTED_BLOCKS);
+            if(tag.isTagged(mainHand) || tag.isTagged(offHand)) return;
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/exoticgarden/listeners/PlantsListener.java:370]: `mainHand` and `offHand` are never assigned.

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]
This is wrong on so many levels.

**Comment by SoSeDiK** [comment]
This seems more like a band-aid than fixing the actual underlying issue 🤔

[BlockStorage.clearBlockInfo(loc);](https://github.com/TheBusyBiscuit/ExoticGarden/blob/master/src/main/java/io/github/thebusybiscuit/exoticgarden/ExoticGarden.java#L422) call adds the location to the removal queue, and then sets the block to AIR.
Could it be that if you spam right click you can place the block after it's gone (set to AIR), but before it's removed from the Slimefun's data storage due to the tick difference?

(maybe there's something even more as to why the gravity affected blocks are even dropped and kept in the first place)

**Comment by SleepyScallop** [comment]
Do what you want then. The patch works just fine on my server.

File: src/main/java/io/github/thebusybiscuit/exoticgarden/listeners/PlantsListener.java (hunk 6 above)
Comment (SleepyScallop) [reply src/main/java/io/github/thebusybiscuit/exoticgarden/listeners/PlantsListener.java:370]: Yesthey are, must have forgot the variable imports.

File: src/main/java/io/github/thebusybiscuit/exoticgarden/listeners/PlantsListener.java (hunk 5 above)
Comment (SleepyScallop) [reply src/main/java/io/github/thebusybiscuit/exoticgarden/listeners/PlantsListener.java:53]: Tell that to the error i was having without static

File: src/main/java/io/github/thebusybiscuit/exoticgarden/listeners/PlantsListener.java (hunk 3 above)
Comment (SleepyScallop) [reply src/main/java/io/github/thebusybiscuit/exoticgarden/listeners/PlantsListener.java:371]: Works for me

**Pull request closed without merging**